### PR TITLE
xwax: Update to Version 1.8 + fixes for lp1956018

### PR DESCRIPTION
**Pull request by daschuer**
Mark has also fixed https://bugs.launchpad.net/mixxx/+bug/1956018 upstream, which is also included.  

All our changes are now tracked via patch files.

Note: This changes the xwax license form GPL 2 to GPL 3

### CODE REVIEW

**Review by Holzhaus — COMMENTED** [review]
Can you elaborate where the patch files are referenced/used?

I suppose you filed upstream PRs. Were they rejected or just not merged yet? Can you add a comment with the appropriate links to the cmake file or something like that?

**Comment by daschuer** [comment]
I have created them by copy our source to the 1.4 branch of xwax. And than I have used Git blame to find the origin of the changes. Finally I have rebased the changes to the master brach of xwax to make them apply cleanly.

**Comment by daschuer** [comment]
I have just send them to the xwax mailinglist.

File: lib/xwax/0005-Fix-traktor-math-issues-for-timecoder-bad-values-lp-.patch [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+diff --git a/timecoder.c b/timecoder.c
+index 63a9a27..da89655 100755
+--- a/timecoder.c
++++ b/timecoder.c
+@@ -640,6 +640,11 @@ signed int timecoder_get_position(struct timecoder *tc, double *when)
+     if (r == -1)
+         return -1;
+ 
++    if (r >= 0) {
```
Comment (ywwg) [inline lib/xwax/0005-Fix-traktor-math-issues-for-timecoder-bad-values-lp-.patch:19]: Mark has told me that we are using his library wrong, and this is probably evidence of that.  This isn't something to fix now, but we should look at it in the future.

File: lib/xwax/timecoder.c [hunk 2]
```diff
@@ -96,7 +96,7 @@ static struct timecode_def timecodes[] = {
         .seed = 0x134503,
         .taps = 0x041040,
         .length = 1500000,
-        .safe = 605000,
+        .safe = 1480000,
```
Comment (ywwg) [inline lib/xwax/timecoder.c:99]: are these changes to the safe value from him or from me?  If he's made the values bigger we might be able to delete our patch.

File: src/vinylcontrol/defs_vinylcontrol.h [hunk 3]
```diff
@@ -8,22 +8,38 @@ constexpr int VINYL_STATUS_OK = 1;
...
-#define MIXXX_VINYL_FINALSCRATCH_LEADIN 0
-#define MIXXX_VINYL_MIXVIBESDVS_LEADIN 0
 #define MIXXX_VINYL_SERATOCV02VINYLSIDEA_LEADIN 28
 #define MIXXX_VINYL_SERATOCV02VINYLSIDEB_LEADIN 28
 #define MIXXX_VINYL_SERATOCD_LEADIN 0
 #define MIXXX_VINYL_TRAKTORSCRATCHSIDEA_LEADIN 10
 #define MIXXX_VINYL_TRAKTORSCRATCHSIDEB_LEADIN 10
+#define MIXXX_VINYL_MIXVIBESDVS_LEADIN 0
+#define MIXXX_VINYL_MIXVIBES7INCH_LEADIN 0
+#define MIXXX_VINYL_PIONEERA_LEADIN 0
```
Comment (ywwg) [inline src/vinylcontrol/defs_vinylcontrol.h:41]: pioneer is modeled on the serato vinyl, so let's set the leadin to 10

**Review by ywwg — CHANGES_REQUESTED** [review]
just a minor change to the leadin value, but otherwise this looks good, thank you!

File: src/vinylcontrol/defs_vinylcontrol.h (hunk 3 above)
Comment (daschuer) [reply src/vinylcontrol/defs_vinylcontrol.h:41]: OK.

File: lib/xwax/timecoder.c (hunk 2 above)
Comment (daschuer) [reply lib/xwax/timecoder.c:99]: From digging through the history Mark has never changed these values. However it looks like I have messed up the patch. 
I will redo it.

Comment (daschuer) [reply lib/xwax/timecoder.c]: Now the patch is fixed. It was a wrong conflict resolving: 

https://github.com/mixxxdj/mixxx/blob/25d17b1ea31873289bea16e6b0c07a75d8d45dc6/lib/xwax/0003-Use-my-the-safe-values-I-found-work-better.patch

Can you remember why you have changed these values originaly?

**Comment by ronso0** [comment]
A related question regarding the lead-in times:
has someone a ressource handy which lists the length of the different vc types so we can implement [lp:1396066 Clamp values for Lead-In times in Preferences> Vinylcontrol](https://bugs.launchpad.net/mixxx/+bug/1396066) ?

File: lib/xwax/timecoder.c (hunk 2 above)
Comment (ywwg) [reply lib/xwax/timecoder.c]: These values determine how much of the vinyl signal is in the very outside of the physical record.  On the records I tested, there's a good ten seconds while the needle is still in the very outer groove, which is where the dj drops the needle. The default values meant that dropping the needle in this part of the record would seek to a few seconds into the track.  I increased the values to make the experience better.

**Comment by ywwg** [comment]
> has someone a ressource handy which lists the length of the different vc types so we can implement [lp:1396066 Clamp values for Lead-In times in Preferences> Vinylcontrol](https://bugs.launchpad.net/mixxx/+bug/1396066) ?

The total usable vinyl length is the ".safe" value in timecoder.c: https://github.com/mixxxdj/mixxx/blob/main/lib/xwax/timecoder.c#L63.  The lead-in should really never be more than 60 seconds though, that's a decent clamp value

**Comment by ywwg** [comment]
(technically there is timecode past the safe value)

**Comment by ywwg** [comment]
Do we need anything else for this?  Mark is eager to get mixxx on this new version so we can resolve the licensing discrepancy

**Comment by Holzhaus** [comment]
Not that I'm away, except some testing maybe?

File: lib/xwax/0005-Fix-traktor-math-issues-for-timecoder-bad-values-lp-.patch (hunk 1 above)
Comment (Holzhaus) [reply lib/xwax/0005-Fix-traktor-math-issues-for-timecoder-bad-values-lp-.patch:19]: maybe leave a TODO comment?

**Comment by ywwg** [comment]
ok I have time to check it tomorrow with serato and traktor v1 vinyl

**Comment by ywwg** [comment]
Tested with serato and traktor V1 vinyl, works as expected

**Comment by Holzhaus** [comment]
Alright, thanks. The PR is still marked as "Changes Requested" due to your earlier review. Please approve it when ypi think it's ready to merge.

**Review by ywwg — APPROVED** [review]

**Comment by ywwg** [comment]
(merge when ready)

**Comment by Holzhaus** [comment]
It's not a draft PR, so this is ready I guess.

**Pull request merged**